In [2]:
import keras
keras.__version__

Using TensorFlow backend.


'2.3.1'

# Objective:  Handwriting Recognition

Dataset: MNIST 

### Step1:  Download/load mnist

In [3]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

__Result:__
1. `train_images`: an array of image pixel data (uint8) 
2. `train_labels`: an array of the corresponding result of the image. 

### Training Dataset

In [6]:
train_images[0].shape   # 28 x 28 matrix

(28, 28)

__Question:__ What's the array/matrix here? 

In [7]:
train_images[0]  

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   3,
         18,  18,  18, 126, 136, 175,  26, 166, 255, 247, 127,   0,   0,
          0,   0],
       [  

__Ans:__
1. It is a 28 x 28 image, which present 5. (How do I know this is 5?)
2. 28 x 28 means, it is a 28 x 28 matrix, which is present in array in numpy.

__Question:__ Why `uint8`?     
__Ans:__     
Because RGB color is presented by [0,255] which is the range of `uint8`.


In [4]:
train_images.shape   # Total images, 60000

(60000, 28, 28)

In [4]:
len(train_labels)   # Total labels, 60000. Certainly, one image present a number, which is called label here. 

60000

__Question: what is 5 here?__      

In [10]:
train_labels   # Total labels.  

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

__Ans:__ 5 means the first image is number 5.  From the 28 x 28 matrix, you would not know what number it is. Therefore, we need the `train_labels` to record the numbers.

### Testing Dataset

In [6]:
test_images.shape

(10000, 28, 28)

In [7]:
len(test_labels)

10000

In [8]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

## Step 2: Build a Neural Network

In [11]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

In [12]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])


## Step3:  Convert train_image and test_images to binary class matrix

In [13]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

We also need to categorically encode the labels, a step which we explain in chapter 3:

In [12]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

### Step4: Train the data

Two quantities are being displayed during training: the "loss" of the network over the training data, and the accuracy of the network over 
the training data.

We quickly reach an accuracy of 0.989 (i.e. 98.9%) on the training data. Now let's check that our model performs well on the test set too:

In [13]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
60000/60000 [==============================] - 2s 32us/step - loss: 0.2570 - accuracy: 0.9250
Epoch 2/5
60000/60000 [==============================] - 2s 31us/step - loss: 0.1028 - accuracy: 0.9695
Epoch 3/5
60000/60000 [==============================] - 2s 31us/step - loss: 0.0682 - accuracy: 0.9795
Epoch 4/5
60000/60000 [==============================] - 2s 32us/step - loss: 0.0492 - accuracy: 0.9853
Epoch 5/5
60000/60000 [==============================] - 2s 32us/step - loss: 0.0375 - accuracy: 0.9890


### Step5: Feed the testing data

In [14]:
test_loss, test_acc = network.evaluate(test_images, test_labels)

10000/10000 [==============================] - 0s 18us/step


In [15]:
print('test_acc:', test_acc)

test_acc: 0.9799000024795532
